src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -225,8 +266,126 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
+        case _ => aggExpr
+      }
+    } flatMap { aggExpr =>
+      aggExpr match {
+        // We have to separate average into sum and count
+        // and for outside expression such as average(x) + 1,
+        // Spark has lift agg + 1 up to resultExpressions
+        // We need to modify the reference there as well to forge
+        // Divide(sum/count) + 1
+        case aggExpr @ AggregateExpression(Average(ref), _, _, _) =>
+          // Need a type promotion
+          val sumToPush = newAggregate(Sum(ref), aggExpr)
+          val countToPush = newAggregate(Count(ref), aggExpr)
+
+          // Need a new expression id since they are not simply rewrite as above
+          val sumFinal = newAggregateWithId(Sum(toAlias(sumToPush).toAttribute), aggExpr)
+          val countFinal = newAggregateWithId(Sum(toAlias(countToPush).toAttribute), aggExpr)
+
+          avgPushdownRewriteMap(aggExpr.resultId) = List(sumToPush, countToPush)
+          avgFinalRewriteMap(aggExpr.resultId) = List(sumFinal, countFinal)
+          List(sumFinal, countFinal)
+        case _ => aggExpr :: Nil
+      }
+    }
+
+    val pushdownAggregates = aggregateExpressions.flatMap { aggExpr =>
+      avgPushdownRewriteMap
+        .getOrElse(aggExpr.resultId, List(aggExpr))
+    }.distinct
+
+    aggregationToDAGRequest(groupingExpressions, pushdownAggregates, source, dagReq)
+
+    val rewrittenResultExpression = resultExpressions.map(
+      expr =>
+        expr
+          .transformDown {
+            case aggExpr: AttributeReference if avgFinalRewriteMap.contains(aggExpr.exprId) =>
+              // Replace the original Average expression with Div of Alias
+              val sumCountPair = avgFinalRewriteMap(aggExpr.exprId)
+
+              // We missed the chance for auto-coerce already
+              // so manual cast needed
+              // Also, convert into resultAttribute since
+              // they are created by tiSpark without Spark conversion
+              // TODO: Is DoubleType a best target type for all?
+              Cast(
+                Divide(
+                  Cast(sumCountPair.head.resultAttribute, DoubleType),
+                  Cast(sumCountPair(1).resultAttribute, DoubleType)
+                ),
+                aggExpr.dataType
+              )
+            case other => other
+          }
+          .asInstanceOf[NamedExpression]
+    )
+
+    val output = (pushdownAggregates.map(x => toAlias(x)) ++ groupingExpressions)
+      .map(_.toAttribute)
+
+    val projectSeq: Seq[Attribute] = projects.asInstanceOf[Seq[Attribute]]
+    projectSeq.foreach(attr => dagReq.addRequiredColumn(TiColumnRef.create(attr.name)))
+
+    aggregate.AggUtils.planAggregateWithoutDistinct(
+      groupingExpressions,
+      residualAggregateExpressions,
+      rewrittenResultExpression,
+      toCoprocessorRDD(source, output, dagReq)
+    )
+  }
+
+  def validGroupAggregate(groupingExpressions: Seq[NamedExpression],
```

Comment:
Change it to isValidAggregates. Aggregates include aggexpr and groupby exprs.

Reply:
Changed.